[AArch64] Remove redundant FMOV for zero-extended i32/i16 loads to f64

Amichaxx opened the pull request:
Previously, a separate load, zext and FMOV instruction was emitted. This patch adds a new TableGen pattern to avoid the unnecessary FMOV. A test is included in test/CodeGen/AArch64/load_u64_from_u32.ll

davemgreen reviewed (commented):
Hi - looks like a good patch. Loads/stores can have a number of addressing modes and types that it might be worth trying to fill in too.

Amichaxx commented on llvm/test/CodeGen/AArch64/load_u64_from_u32.ll:
```diff
@@ -0,0 +1,14 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=aarch64-linux-gnu -o - %s | FileCheck %s
+
+define double @_Z9load_u64_from_u32_testPj(ptr %n) {
```
Yes, will get to adding these today. Thanks.

Amichaxx commented on llvm/lib/Target/AArch64/AArch64InstrInfo.td:
```diff
@@ -9414,6 +9418,7 @@ def : Pat<(v4i32 (mulhu V128:$Rn, V128:$Rm)),
                              (EXTRACT_SUBREG V128:$Rm, dsub)),
            (UMULLv4i32_v2i64 V128:$Rn, V128:$Rm))>;
 
+
```
Changed in latest commit.

Amichaxx committed:
[AArch64] Remove redundant FMOV for zero-extended i32/i16 loads to f64

Previously, a load from i32 or i16, followed by zero-extension to i64 and a
bitcast to f64, would emit separate FMOV instructions.

This patch introduces new corresponding TableGen patterns to avoid the unnecessary FMOV.

Tests added:
  - load_u64_from_u32.ll
  - load_u64_from_u16.ll

Amichaxx committed:
[AArch64] This patch removes redundant FMOV instructions for patterns where an integer is loaded from memory, zero-extended, and bitcast to a floating-point type. Patterns are added for i8/i16/i32 to f64 and f32. A single test file (load-zext-bitcast.ll) is included for all cases, which combines the previously created test files.

davemgreen reviewed (approved):
LGTM, thanks.

davemgreen commented on llvm/lib/Target/AArch64/AArch64InstrInfo.td:
```diff
@@ -10986,4 +11006,4 @@ defm FMMLA : SIMDThreeSameVectorFP8MatrixMul<"fmmla">;
 include "AArch64SVEInstrInfo.td"
 include "AArch64SMEInstrInfo.td"
-include "AArch64InstrGISel.td"
+include "AArch64InstrGISel.td"
```
Hi - I you update this change we can get this submitted. Thanks.

Amichaxx replied:
Apologies, I didn't see your comment. It should be rectified now.

davemgreen reviewed (approved):
Thank you. LGTM
